Subarray introduction
total no. of subarrays = n +(n-1) + (n-2) + ..+3+2+1 = n*(n+1)/2

In [75]:
A = [1,2,3,4,5]
class SolutionSubarrayIntro:
    def __init__(self,array_p):
        self.A = array_p

    def subarray_print(self,start,end):
        ans =[]
        for i in range(start,end+1):
            ans.append(self.A[i])
        return ans
    
    def subarray_all_print(self):
        n = len(self.A)
        ans =[]
        for i in range(n): # start   
            for j in range(i,n): # end 
                row=[]
                for k in range(i,j+1):
                    row.append(A[k])
                ans.append(row)
        return ans
    
    def subarray_all_print_newline(self):
        n = len(self.A)
        for i in range(n):
            for j in range(i,n):
                #row=[]
                for k in range(i,j+1):
                    #row.append(A[k])
                    print(A[k],end=" ")
                print()
    def subarray_sum_bruteforce(self):
        n = len(self.A)
        
        for i in range(n):# start index 
            for j in range(i,n): # end index
                sum_s=0
                for k in range(i,j+1):
                    sum_s+= A[k]
                print(sum_s)

    def subarray_sum_prefixsum(self):
        def prefix_sum():
            n = len(self.A)
            sum_p = 0
            P = []
            for i in range(n):
                sum_p += self.A[i]
                P.append(sum_p)
            return P
        P = prefix_sum()
        n = len(self.A)
        for i in range(n):
            for j in range(i,n):
                if i == 0:
                    print(P[j])
                else:
                    print(P[j]-P[i-1])

    def subarray_sum_carryforward(self,start,end):
        #n = len(self.A)
        sum_c = 0
        for i in range(start,end+1):
            sum_c += self.A[i]
            print(sum_c)

    
    def subarray_all_sum_carryforward(self):
        n = len(self.A)
        for i in range(n):# i = start index 
            self.subarray_sum_carryforward(i,n-1)

    def subarray_all_sum_carryforward1(self):
        n = len(self.A)
        ans =[]
        for i in range(n):
            sum_c = 0
            for j in range(i,n):
                sum_c += self.A[j]
                ans.append(sum_c)
                #print(sum_c)
        return ans
    
    def subarray_maxsum_carryforward(self):
        n = len(self.A)
        ans = float('-inf')
        for i in range(n):
            sum_c = 0
            for j in range(i,n):
                sum_c += self.A[j]
                ans = max(ans,sum_c)

        return ans
    
    def subarray_minsum_carryforward(self):
        n = len(self.A)
        ans = float('+inf')
        for i in range(n):
            sum_c = 0
            for j in range(i,n):
                sum_c += self.A[j]
                ans = min(ans,sum_c)

        return ans
    
    def subarray_evensum_carryforward(self):
        n = len(self.A)
        count = 0
        for i in range(n):
            sum_c = 0
            for j in range(i,n):
                sum_c += self.A[j]
                if (sum_c % 2 == 0):
                    count += 1
                #print(sum_c)
        return count
    
    def subarray_oddsum_carryforward(self):
        n = len(self.A)
        count = 0
        for i in range(n):
            sum_c = 0
            for j in range(i,n):
                sum_c += self.A[j]
                if (sum_c % 2 != 0):
                    count += 1
                #print(sum_c)
        return count
    
    def subarray_evenlengthsum_gt_B_carryforward(self,B):
        n = len(self.A)
        count = 0
        for i in range(n):
            sum_c = 0
            for j in range(i,n):
                sum_c += self.A[j]
                length = j-i + 1
                if (length % 2 == 0 and sum_c > B):
                    count += 1
                #print(sum_c)
        return count
    
    def subarray_all_globalsum_carryforward(self):
        n = len(self.A)
        ans =[]
        for i in range(n):
            sum_c = 0
            for j in range(i,n):
                sum_c += self.A[j]
                ans.append(sum_c)
                #print(sum_c)
        return sum(ans)
    
    def subarray_all_globalsum_carryforward1(self):
        n = len(self.A)
        gsum=0
        for i in range(n):
            sum_c = 0
            for j in range(i,n):
                sum_c += self.A[j]
                gsum += sum_c
                #print(sum_c)
        return gsum
    
    def subarray_all_globalsum_prefixsum(self):
        def prefix_sum(array):
            n = len(array)
            sum_p = 0
            P = []
            for i in range(n):
                sum_p += array[i]
                P.append(sum_p)
            return P
        P = prefix_sum(self.A)
        n = len(self.A)
        P1 =[]
        for i in range(n):
            for j in range(i,n):
                if i == 0:
                    P1.append(P[j])
                else:
                    P1.append(P[j]-P[i-1])
        P2 = prefix_sum(P1)
        return P2[-1]
    
    def subarray_globalsum_optimized(self):
        gsum = 0
        n = len(self.A)
        for i in range(n):
            gsum += self.A[i]*(n-i)*(i+1)
        return gsum
    
    def subarray_index_occurrence(self,index):
        n = len(self.A)
        return (n-index)*(index+1)









    
s1 = SolutionSubarrayIntro(A)
#s1.subarray_print(2,4)
#s1.subarray_all_print()
s1.subarray_all_print_newline()
#s1.subarray_sum_bruteforce()
#print("Check")
#s1.subarray_sum_prefixsum()
#s1.subarray_sum_carryforward(2,4)
#s1.subarray_all_sum_carryforward()
#s1.subarray_all_sum_carryforward1()
#s1.subarray_maxsum_carryforward()
#s1.subarray_minsum_carryforward()
#s1.subarray_evensum_carryforward()
#s1.subarray_oddsum_carryforward()
#s1.subarray_evenlengthsum_gt_B_carryforward(3)
#s1.subarray_all_globalsum_carryforward()
#s1.subarray_all_globalsum_prefixsum()
#s1.subarray_all_globalsum_carryforward1()
s1.subarray_globalsum_optimized()
s1.subarray_index_occurrence(3)


1 
1 2 
1 2 3 
1 2 3 4 
1 2 3 4 5 
2 
2 3 
2 3 4 
2 3 4 5 
3 
3 4 
3 4 5 
4 
4 5 
5 


8

In [46]:
A=[1,2,3,4,5]
n=len(A)
c=[]
for i in range(n):
    c.append((n-i)*(i+1))
sum_a = 0
for i in range(len(A)):
    sum_a += A[i]*(n-i)*(i+1)
print(sum_a)


105


total no. of Subarrays of len k = n-k+1
Total no. of occurrence of any index i (0 based ) in an array of length n with total n*(n+1)/2 = (n-i)*(i+1)

In [112]:
A = [1,2,3,4,5]
A =[-7,3,4,-2,5,3,-20,8,2,-10,4]
class SolutionSubarrayAdv:
    def __init__(self,array_p):
        self.A = array_p

    def subarray_maxsum_len(self,k):#sliding window
        n = len(self.A)
        ans=float('-inf')
        sum_k = 0
        for i in range(k):
            sum_k += self.A[i]
        ans = max(ans,sum_k)
        for j in range(1,n-k+1):
            sum_k = sum_k - self.A[j-1] + self.A[j+k-1]
            ans= max(ans,sum_k)
        return ans
    
    def subarray_maxsum_bruteforce(self):
        n = len(self.A)
        ans = float('-inf')
        for start in range(n):
            for end in range(start,n):
                sum_m = 0
                for i in range(start,end+1):
                    sum_m+=self.A[i]
                if sum_m > ans:
                    ans = sum_m
        return ans
    
    def subarray_maxsum_kadane_algo(self):
        sum_k = 0
        ans = float('-inf') # mas sum initiation
        for i,x in enumerate(self.A):
            sum_k += x
            # the below line can be written as : ans = max(sum_k,ans)
            if sum_k > ans:
                ans = sum_k
            # the below line can be writtne as : sum_k = max(sum_k,0)
            if sum_k < 0:
                sum_k =0
        return ans
    
    def subarray_maxsum_kadane_algo_with_indices(self):
        sum_k = 0
        ans = float('-inf')
        final = []
        best_start= 0
        best_end = 0
        flagS = 0
        flagP = 0
        flagN = 0
        state =[]
        for i,x in enumerate(self.A):
            sum_k+=x
            if sum_k < ans:
                flagS=1
                final.append(f"sustainTrend {i} {ans} {sum_k}")
            if sum_k > ans:
                ans = sum_k
                if i == 0:
                    best_start = i
                    flagP = 2
                else:
                    best_end = i
                    flagP=2

                final.append(f"posTrend {i} {ans} {sum_k}")
            if sum_k <0 :
                sum_k = 0
                flagN = 3
                
                final.append(f"negTrend {i} {ans} {sum_k}")

            
            state.append([flagS,flagP,flagN])
            #print(f"possible {best_start} {best_end}")
        return final,state,ans

    # pos > neg > pos then take latest pos.
    # pos > neg > sus > neg then take the pos
    # pos > neg > sus > neg > pos then take the latest pos
    # pos > neg then take the pos
    # pos > neg > sus then take the pos
    def subarray_alternative_print_all(self):
        pass

    def subarray_fromright_print_all(self):
        pass

A =[-7,3,4,-2,5,3,-20,8,2,-10,4]
A=[-7,8]
A=[-7,-2]
A=[-7,-9]
A=[-7]
A=[-7,-2,-2,-1,-4,-3]
A =[-7,-8,3,4,-2,5,3,-20,8,2,-10,4]
A = [2,3,5,8]
A=[-4,-6,-2,-7,-3,-10]
"""
-4
1. -2, +5
> 2 , 8
1.1. -1, 3 or 0
1.2. -6 
2. -7
"""
#A=[-7,-8,-3]
s2 = SolutionSubarrayAdv(A)
#s2.subarray_maxsum_len(3)
#s2.subarray_maxsum_bruteforce()
#s2.subarray_maxsum_kadane_algo()
s2.subarray_maxsum_kadane_algo_with_indices()

(['posTrend 0 -4 -4',
  'negTrend 0 -4 0',
  'sustainTrend 1 -4 -6',
  'negTrend 1 -4 0',
  'posTrend 2 -2 -2',
  'negTrend 2 -2 0',
  'sustainTrend 3 -2 -7',
  'negTrend 3 -2 0',
  'sustainTrend 4 -2 -3',
  'negTrend 4 -2 0',
  'sustainTrend 5 -2 -10',
  'negTrend 5 -2 0'],
 [[0, 2, 3], [1, 2, 3], [1, 2, 3], [1, 2, 3], [1, 2, 3], [1, 2, 3]],
 -2)

In [77]:
A =[3,4,-2,5,3,-20,8,2,-10,4]
for i,x in enumerate(A):
    print(i,x)

0 3
1 4
2 -2
3 5
4 3
5 -20
6 8
7 2
8 -10
9 4


Q.3. Prefix maximum
Unsolved
feature icon
Get your doubts resolved blazing fast with Chat GPT Help
Check Chat GPT
feature icon
Using hints is now penalty free
Use Hint
Problem Description
Kamal is a software developer and he's working on a new feature for an e-commerce website. The website has a list of prices for different products, and Kamal needs to find the maximum price of all products up to a given index.

He has the list of prices in an array A of length N, and he needs to write a program that will return the maximum price occurring in the subarray from 0 to i for every index i. Kamal needs your help to implement this function.



Problem Constraints
1 <= N <= 105
1 <= A[i] <= 109


Input Format
Only argument A is an array of integers.


Output Format
Return an array of integers where the i-th element of the array represents the maximum value occurring in the subarray from 0 to i in the input array A.


Example Input
Input 1:

A = [9, 7, 6, 18, 2]
Input 2:

A = [16, 8, 24, 9, 25, 17]


Example Output
Output 1:

[9, 9, 9, 18, 18]
Output 2:

[16, 16, 24, 24, 25, 25]


Example Explanation
Explanation 1:

The prefix maximum array of [9, 7, 6, 18, 2] is [9, 9, 9, 18, 18].
Explanation 2:

The prefix maximum array of [16, 8, 24, 9, 25, 17] is [16, 16, 24, 24, 25, 25].

In [53]:
class SolutionPrefixMax:
    def __init__(self,array_p):
        self.A = array_p

    def prefix_max(self):
        n = len(self.A)
        P = []
        P.append(self.A[0])
        for i in range(1,n):
            P.append(max(P[i-1],self.A[i]))
        return P
    
A = [9, 7, 6, 18, 2]
s3 = SolutionPrefixMax(A)
s3.prefix_max()

[9, 9, 9, 18, 18]

Q.4. Suffix maximum
Unsolved
feature icon
Get your doubts resolved blazing fast with Chat GPT Help
Check Chat GPT
feature icon
Using hints is now penalty free
Use Hint
Problem Description
Given an array A of length N, For every index i, you need to find the maximum value in subarray from i to N-1.


Problem Constraints
1 <= N <= 105
1 <= A[i] <= 109


Input Format
Only argument A is an array of integers.


Output Format
Return an array of integers denoting the suffix maximum of the given array.


Example Input
Input 1:

A = [12, 5, 6, 7]
Input 2:

A = [15, 9, 7, 11, 10]


Example Output
Output 1:

[12, 7, 7, 7]
Output 2:

[15, 11, 11, 11, 10]


Example Explanation
Explanation 1:

The suffix maximum array of [12, 5, 6, 7]] is [12, 7, 7, 7].
Explanation 2:

The suffix maximum array of [15, 9, 7, 11, 10] is [15, 11, 11, 11, 10].

In [60]:
class SolutionSuffixMax:
    def __init__(self,array_p):
        self.A = array_p

    def suffix_max(self):
        n = len(self.A)
        S=[]
        S.append(self.A[n-1])
        for i in reversed(range(n-1)):
            S.append(max(S[n-2-i],self.A[i]))
        return S[::-1]

A = [12, 5, 6, 7]       
s4 = SolutionSuffixMax(A)
s4.suffix_max()

[12, 7, 7, 7]

Q.5. Continuous Sum Query
Unsolved
feature icon
Get your doubts resolved blazing fast with Chat GPT Help
Check Chat GPT
feature icon
Using hints is now penalty free
Use Hint
Problem Description
There are A beggars sitting in a row outside a temple. Each beggar initially has an empty pot. When the devotees come to the temple, they donate some amount of coins to these beggars. Each devotee gives a fixed amount of coin(according to their faith and ability) to some K beggars sitting next to each other.

Given the amount P donated by each devotee to the beggars ranging from L to R index, where 1 <= L <= R <= A, find out the final amount of money in each beggar's pot at the end of the day, provided they don't fill their pots by any other means.
For ith devotee B[i][0] = L, B[i][1] = R, B[i][2] = P, Given by the 2D array B


Problem Constraints
1 <= A <= 2 * 105
1 <= L <= R <= A
1 <= P <= 103
0 <= len(B) <= 105


Input Format
The first argument is a single integer A.
The second argument is a 2D integer array B.


Output Format
Return an array(0 based indexing) that stores the total number of coins in each beggars pot.


Example Input
Input 1:-
A = 5
B = [[1, 2, 10], [2, 3, 20], [2, 5, 25]]


Example Output
Output 1:-
10 55 45 25 25


Example Explanation
Explanation 1:-
First devotee donated 10 coins to beggars ranging from 1 to 2. Final amount in each beggars pot after first devotee: [10, 10, 0, 0, 0]
Second devotee donated 20 coins to beggars ranging from 2 to 3. Final amount in each beggars pot after second devotee: [10, 30, 20, 0, 0]
Third devotee donated 25 coins to beggars ranging from 2 to 5. Final amount in each beggars pot after third devotee: [10, 55, 45, 25, 25]

In [71]:
A = 5
P = [0 for i in range(A)]
B = [[1, 2, 10], [2, 3, 20], [2, 5, 25]]
#for i in range(A):
print(P)
'''
for left,right,value in B:
    print(left-1,right-1,value)
'''

for left,right,value in B:
    print(left-1,right-1,value)
    



[0, 0, 0, 0, 0]
0 1 10
1 2 20
1 4 25
